File: backend/app/api/v1/intelligence.py
from datetime import datetime, timedelta, timezone
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.config import settings
from app.core.database import get_db
from app.crud import child as crud_child
from app.models.feedback import DailyFeedback
from app.models.session import Session as SessionModel
from app.models.session_notes import SessionNotes
from app.models.therapist import TherapistProfile
from app.models.user import User
from app.schemas.intelligence import (
    AutoTagResponse,
    MatchingResponse,
    NarrativeReportResponse,
    SkillScoreResponse,
    TherapistMatch,
)
from app.services import intelligence as svc

router = APIRouter()


# ── AI-001: Skill Score ───────────────────────────────────────────────────────

@router.get("/skill-score/{child_id}", response_model=SkillScoreResponse)
async def get_skill_score(
    child_id: str,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    child = crud_child.get_child_by_id(db, child_id)
    if not child:
        raise HTTPException(status_code=404, detail="Child not found")

    session_ids = [
        s.id
        for s in db.query(SessionModel).filter(SessionModel.child_id == child_id).all()
    ]
    notes = (
        db.query(SessionNotes)
        .filter(SessionNotes.session_id.in_(session_ids))
        .all()
        if session_ids else []
    )

    result = svc.compute_skill_score(child, notes)
    return SkillScoreResponse(child_id=child_id, child_name=child.name, **result)


# ── AI-003: Therapist Matching ────────────────────────────────────────────────

@router.get("/match-therapist/{child_id}", response_model=MatchingResponse)
async def match_therapist(
    child_id: str,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    child = crud_child.get_child_by_id(db, child_id)
    if not child:
        raise HTTPException(status_code=404, detail="Child not found")

    profiles = (
        db.query(TherapistProfile, User)
        .join(User, User.id == TherapistProfile.user_id)
        .filter(User.role == "therapist")
        .all()
    )

    matches = svc.match_therapists(child, profiles)
    return MatchingResponse(
        child_id=child_id,
        child_name=child.name,
        diagnosis=child.diagnosis,
        therapy_goals=child.therapy_goals or [],
        matches=[TherapistMatch(**m) for m in matches],
    )


# ── AI-002: Auto-tag Session Notes ───────────────────────────────────────────

@router.post("/auto-tag/{session_id}", response_model=AutoTagResponse)
async def auto_tag_session(
    session_id: str,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    notes = (
        db.query(SessionNotes)
        .filter(SessionNotes.session_id == session_id)
        .first()
    )
    if not notes:
        raise HTTPException(status_code=404, detail="Session notes not found")
    if not notes.observations:
        return AutoTagResponse(session_id=session_id, tags=[])

    tags = await svc.auto_tag_observations(notes.observations, settings.ANTHROPIC_API_KEY)
    return AutoTagResponse(session_id=session_id, tags=tags)


# ── AI-004: Narrative Report Generator ───────────────────────────────────────

@router.get("/generate-report/{child_id}", response_model=NarrativeReportResponse)
async def generate_report(
    child_id: str,
    period: Literal["weekly", "monthly"] = "weekly",
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    child = crud_child.get_child_by_id(db, child_id)
    if not child:
        raise HTTPException(status_code=404, detail="Child not found")

    days = 7 if period == "weekly" else 30
    since = datetime.now(timezone.utc) - timedelta(days=days)

    sessions = (
        db.query(SessionModel)
        .filter(SessionModel.child_id == child_id, SessionModel.scheduled_at >= since)
        .all()
    )
    session_ids = [s.id for s in sessions]
    notes = (
        db.query(SessionNotes).filter(SessionNotes.session_id.in_(session_ids)).all()
        if session_ids else []
    )
    feedbacks = (
        db.query(DailyFeedback)
        .filter(
            DailyFeedback.child_id == child_id,
            DailyFeedback.feedback_date >= since.date(),
        )
        .all()
    )

    completed = sum(1 for s in sessions if s.status.value == "completed")

    score_fields = [
        "attention_score", "participation_score",
        "mood_score", "progress_score", "behavior_score",
    ]
    dim_avgs: dict[str, float] = {}
    for field in score_fields:
        vals = [getattr(n, field) for n in notes if getattr(n, field) is not None]
        dim_avgs[field] = round(sum(vals) / len(vals), 2) if vals else 0.0

    avg_progress = dim_avgs.get("progress_score", 0.0)
    non_zero = {k: v for k, v in dim_avgs.items() if v > 0}
    strongest = max(non_zero, key=non_zero.get) if non_zero else "N/A"
    weakest = min(non_zero, key=non_zero.get) if non_zero else "N/A"

    goals = list({g for n in notes for g in (n.goals_worked_on or [])})[:5]
    latest_note = notes[0].next_session_plan if notes else None

    fb_mood = [f.mood_score for f in feedbacks if f.mood_score is not None]
    avg_feedback_mood = round(sum(fb_mood) / len(fb_mood), 1) if fb_mood else 0.0

    data = {
        "child_name": child.name,
        "sessions_completed": completed,
        "avg_progress": avg_progress,
        "strongest": strongest.replace("_score", "").replace("_", " "),
        "weakest": weakest.replace("_score", "").replace("_", " "),
        "goals": goals,
        "latest_note": latest_note,
        "avg_feedback_mood": avg_feedback_mood,
    }

    narrative = await svc.generate_narrative_report(data, period, settings.ANTHROPIC_API_KEY)
    return NarrativeReportResponse(
        child_id=child_id,
        child_name=child.name,
        period=period,
        narrative=narrative,
    )

File: backend-node/src/routes/intelligence.js
const express = require('express');
const db = require('../db');
const config = require('../config/config');
const { authenticate } = require('../middleware/auth');
const svc = require('../services/intelligence');

const router = express.Router();

// AI-001: Skill Score
router.get('/skill-score/:childId', authenticate, async (req, res) => {
  const { childId } = req.params;
  const childRes = await db.query('SELECT * FROM children WHERE id = $1', [childId]);
  if (!childRes.rows.length) return res.status(404).json({ detail: 'Child not found' });
  const child = childRes.rows[0];

  const sessionRes = await db.query(
    'SELECT id FROM sessions WHERE child_id = $1', [childId]
  );
  const sessionIds = sessionRes.rows.map(r => r.id);

  let notes = [];
  if (sessionIds.length) {
    const notesRes = await db.query(
      'SELECT * FROM session_notes WHERE session_id = ANY($1)', [sessionIds]
    );
    notes = notesRes.rows;
  }

  const result = svc.computeSkillScore(child, notes);
  res.json({ child_id: childId, child_name: child.name, ...result });
});

// AI-003: Therapist Matching
router.get('/match-therapist/:childId', authenticate, async (req, res) => {
  const { childId } = req.params;
  const childRes = await db.query('SELECT * FROM children WHERE id = $1', [childId]);
  if (!childRes.rows.length) return res.status(404).json({ detail: 'Child not found' });
  const child = childRes.rows[0];

  const profilesRes = await db.query(
    `SELECT tp.*, u.id as user_id, u.name as user_name, u.role as user_role
     FROM therapist_profiles tp
     JOIN users u ON u.id = tp.user_id
     WHERE u.role = 'therapist'`
  );
  const profiles = profilesRes.rows.map(row => ({
    profile: {
      specialization: row.specialization,
      years_of_experience: row.years_of_experience,
      qualification: row.qualification,
      rating: row.rating,
      total_sessions: row.total_sessions,
    },
    user: { id: row.user_id, name: row.user_name },
  }));

  const matches = svc.matchTherapists(child, profiles);
  res.json({
    child_id: childId,
    child_name: child.name,
    diagnosis: child.diagnosis,
    therapy_goals: child.therapy_goals || [],
    matches,
  });
});

// AI-002: Auto-tag Session Notes
router.post('/auto-tag/:sessionId', authenticate, async (req, res) => {
  const { rows } = await db.query(
    'SELECT * FROM session_notes WHERE session_id = $1', [req.params.sessionId]
  );
  if (!rows.length) return res.status(404).json({ detail: 'Session notes not found' });
  const notes = rows[0];
  if (!notes.observations) return res.json({ session_id: req.params.sessionId, tags: [] });

  const tags = await svc.autoTagObservations(notes.observations, config.anthropicApiKey);
  res.json({ session_id: req.params.sessionId, tags });
});

// AI-004: Narrative Report
router.get('/generate-report/:childId', authenticate, async (req, res) => {
  const { childId } = req.params;
  const period = ['weekly', 'monthly'].includes(req.query.period) ? req.query.period : 'weekly';
  const days = period === 'weekly' ? 7 : 30;
  const since = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();

  const childRes = await db.query('SELECT * FROM children WHERE id = $1', [childId]);
  if (!childRes.rows.length) return res.status(404).json({ detail: 'Child not found' });
  const child = childRes.rows[0];

  const sessionsRes = await db.query(
    'SELECT * FROM sessions WHERE child_id = $1 AND scheduled_at >= $2', [childId, since]
  );
  const sessions = sessionsRes.rows;
  const sessionIds = sessions.map(s => s.id);

  let notes = [];
  if (sessionIds.length) {
    const notesRes = await db.query(
      'SELECT * FROM session_notes WHERE session_id = ANY($1)', [sessionIds]
    );
    notes = notesRes.rows;
  }

  const sinceDate = since.slice(0, 10);
  const feedbackRes = await db.query(
    `SELECT * FROM daily_feedback WHERE child_id = $1 AND feedback_date >= $2`,
    [childId, sinceDate]
  );
  const feedbacks = feedbackRes.rows;

  const completed = sessions.filter(s => s.status === 'completed').length;
  const scoreFields = ['attention_score', 'participation_score', 'mood_score', 'progress_score', 'behavior_score'];
  const dimAvgs = {};
  for (const field of scoreFields) {
    const vals = notes.map(n => n[field]).filter(v => v != null);
    dimAvgs[field] = vals.length
      ? Math.round(vals.reduce((a, b) => a + b, 0) / vals.length * 100) / 100 : 0;
  }

  const nonZero = Object.fromEntries(Object.entries(dimAvgs).filter(([, v]) => v > 0));
  const strongest = Object.keys(nonZero).length
    ? Object.entries(nonZero).sort((a, b) => b[1] - a[1])[0][0] : 'N/A';
  const weakest = Object.keys(nonZero).length
    ? Object.entries(nonZero).sort((a, b) => a[1] - b[1])[0][0] : 'N/A';

  const goals = [...new Set(notes.flatMap(n => n.goals_worked_on || []))].slice(0, 5);
  const latestNote = notes.length ? notes[0].next_session_plan : null;
  const fbMood = feedbacks.map(f => f.mood_score).filter(v => v != null);
  const avgFeedbackMood = fbMood.length
    ? Math.round(fbMood.reduce((a, b) => a + b, 0) / fbMood.length * 10) / 10 : 0;

  const data = {
    child_name: child.name,
    sessions_completed: completed,
    avg_progress: dimAvgs['progress_score'] || 0,
    strongest: strongest.replace(/_score$/, '').replace(/_/g, ' '),
    weakest: weakest.replace(/_score$/, '').replace(/_/g, ' '),
    goals,
    latest_note: latestNote,
    avg_feedback_mood: avgFeedbackMood,
  };

  const narrative = await svc.generateNarrativeReport(data, period, config.anthropicApiKey);
  res.json({ child_id: childId, child_name: child.name, period, narrative });
});

module.exports = router;